### PR TITLE
[New] add `no-object-type-as-default-prop` rule

### PR DESCRIPTION
**Pull request by cyan33**
Discussed in 

https://github.com/yannickcr/eslint-plugin-react/issues/2847

## Test

```
npm run lint
npm run test
```

shows no errors and warnings

---

- [x] Added it in README

### CODE REVIEW

**Comment by cyan33** [comment]
cc @ljharb

File: docs/rules/no-reference-type-as-default-prop.md [hunk 1]
```diff
@@ -0,0 +1,68 @@
+# Prevent usage of referential-type variables as default param in functional component (react/no-reference-type-as-default-prop)
+
+Warns if in a functional component, a referential type value (such as array/object literal) is used as default prop, to prevent potential un-necessary re-renders, and performance regressions.
```
Comment (ljharb) [inline docs/rules/no-reference-type-as-default-prop.md:3]: i'm not sure if the name "referential type" is helpful - the hazard is with any object, a category which arrays and functions belong to, so `no-object-as-default-prop` might be clearer? (especially since any implication that JS is a pass-by-reference language would be incorrect and confusing)

File: docs/rules/no-reference-type-as-default-prop.md [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+
+## Rule Details
+
+Certain values (like arrays, objects, functions, etc) are compared by reference instead of by value. This means that, for example, whilst two empty arrays represent the same value - to the JS engine they are distinct and inequal as they represent references to different things in memory.
```
Comment (ljharb) [inline docs/rules/no-reference-type-as-default-prop.md:7]: I'd prefer to describe this as comparing by identity - and note that this also includes Symbols. "reference" is an overloaded term.

File: docs/rules/no-reference-type-as-default-prop.md [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+
+When using object destructuring syntax you can set the default value for a given property if it does not exist. If you set the default value to one of the values that is compared by reference, it will mean that each time the destructure is executed the JS engine will create a new, non-equal value in the destructured variable.
+
+In the  context of a React functional component's props argument this means that each render the property has a new, distinct value. When this value is passed to a hook as a dependency or passed into a child component as a property react will see this as a new value - meaning that a hook will be re-evaluated, or a memoized component will rerender.
+
+This obviously destroys any performance benefits you get from memoisation. Additionally in certain circumstances this can cause infinite rerender loops, which can often be hard to debug.
```
Comment (ljharb) [inline docs/rules/no-reference-type-as-default-prop.md:13]:
```suggestion
This obviously destroys any performance benefits you get from memoization. Additionally in certain circumstances this can cause infinite rerender loops, which can often be hard to debug.
```

File: tests/lib/rules/no-reference-type-as-default-prop.js [hunk 4]
```diff
@@ -0,0 +1,212 @@
...
+    }
+  `,
+    `
+    const Foo = () => {};
+  `,
+    `
+    const Foo = ({bar = 1}) => {};
+  `
+  ],
+  invalid: [
```
Comment (ljharb) [inline tests/lib/rules/no-reference-type-as-default-prop.js:66]: let's make sure each of these test cases is duplicated for the default parser, babel-eslint, and the typescript parsers.

**Review by ljharb — CHANGES_REQUESTED** [review]
LGTM minus the naming/prose concerns, the question about new Symbols, and the test parser comment

**Comment by AriPerkkio** [comment]
This new rule looked useful so I quickly tested it against +150 repositories. It seems to be [crashing quite often](https://github.com/AriPerkkio/eslint-plugin-react/blob/test/no-reference-type-as-default-prop/eslint-remote-tester-results/baseweb.md) on `isReactComponentName` method. This test should cover at least one of those cases
```js
export default function() {
    return <div />;
}
```

I'm not sure if all crashes were caused by the same issue. You can find all results from [here](https://github.com/AriPerkkio/eslint-plugin-react/tree/test/no-reference-type-as-default-prop/eslint-remote-tester-results) and log containing all crashing cases from [here](https://github.com/AriPerkkio/eslint-plugin-react/blob/test/no-reference-type-as-default-prop/lint-log).

**Comment by ljharb** [comment]
@cyan33 ping :-) there's a few things to address

File: docs/rules/no-reference-type-as-default-prop.md (hunk 1 above)
Comment (cyan33) [reply docs/rules/no-reference-type-as-default-prop.md:3]: Sounds good. If we see `object` as a broad term, the indication does seem clearer with the proposed name `no-object-as-default-prop`.

**Comment by cyan33** [comment]
Sorry it took me a bit to get back to this 🙂

Updated:

- Tweaking the name and docs of the rule
- Adding support to check for `Symbol('foo')`
- Add test case for anonymous function
  - Simply skip as it's hard to guarantee if it's a real React component just based on its content. I think some false positive is fine in this case since it's not encouraged to not provide a name to the component?

**Comment by cyan33** [comment]
Ping @ljharb

File: docs/rules/no-object-type-as-default-prop.md [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+
+When using object destructuring syntax you can set the default value for a given property if it does not exist. If you set the default value to one of the values that is compared by identity, it will mean that each time the destructure is executed the JS engine will create a new, non-equal value in the destructured variable.
+
+In the  context of a React functional component's props argument this means that each render the property has a new, distinct value. When this value is passed to a hook as a dependency or passed into a child component as a property react will see this as a new value - meaning that a hook will be re-evaluated, or a memoized component will rerender.
+
+This obviously destroys any performance benefits you get from memoisation. Additionally in certain circumstances this can cause infinite rerender loops, which can often be hard to debug.
+
+It's worth noting that primitive literal values (`string`, `number`, `boolean`, `null`, and `undefined`) are equal when two same value are being compared, so it's safe for those to be inlined as default prop.
```
Comment (ljharb) [inline docs/rules/no-object-type-as-default-prop.md:15]:
```suggestion
It's worth noting that primitive literal values (`string`, `number`, `boolean`, `null`, and `undefined`) can be considered to be compared "by value", or alternatively, as always having the same identity (every `3` is the same exact `3`). Thus, it's safe for those to be inlined as a default value.
```

File: docs/rules/no-object-type-as-default-prop.md [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+
+To fix the violations, the easiest way is to use a referencing variable instead of using the literal values, e.g:
```
Comment (ljharb) [inline docs/rules/no-object-type-as-default-prop.md:17]:
```suggestion
To fix the violations, the easiest way is to use a referencing variable in module scope instead of using the literal values, e.g:
```

File: docs/rules/no-object-type-as-default-prop.md [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+const Component = ({
+  items = emptyObject,
+}) => {}
+
+const noopFunc = () => {};
+const Component = ({
+  items = noopFunc,
+}) => {}
+
+// primitive literals are all compared by value, so are safe to be inlined
```
Comment (ljharb) [inline docs/rules/no-object-type-as-default-prop.md:62]:
```suggestion
// primitives are all compared by value, so are safe to be inlined
```

File: lib/rules/no-object-type-as-default-prop.js [hunk 8]
```diff
@@ -0,0 +1,150 @@
...
+
+const FORBIDDEN_TYPES = new Set(Object.keys(FORBIDDEN_TYPES_MAP));
+const MESSAGE_ID = 'forbiddenTypeDefaultParam';
+
+function isReactComponentName(node) {
+  if (node.id && node.id.type === 'Identifier' && node.id.name) {
+    const firstLetter = node.id.name[0];
+    if (firstLetter.toUpperCase() === firstLetter) {
+      return true;
+    }
```
Comment (ljharb) [inline lib/rules/no-object-type-as-default-prop.js:28]:
```suggestion
    return firstLetter.toUpperCase() === firstLetter;
```

File: lib/rules/no-object-type-as-default-prop.js [hunk 9]
```diff
@@ -0,0 +1,150 @@
...
+
+function hasUsedObjectDestructuringSyntax(params) {
+  if (
+    params == null
+    || params.length !== 1
+    || params[0].type !== 'ObjectPattern'
+  ) {
+    return false;
+  }
+  return true;
```
Comment (ljharb) [inline lib/rules/no-object-type-as-default-prop.js:52]:
```suggestion
  return
    params != null
    && params.length === 1
    && params[0].type === 'ObjectPattern';
```

File: lib/rules/no-object-type-as-default-prop.js [hunk 10]
```diff
@@ -0,0 +1,150 @@
...
+}
+
+function verifyDefaultPropsDestructuring(context, properties) {
+  // Loop through each of the default params
+  properties.forEach((prop) => {
+    if (prop.type !== 'Property') {
+      return;
+    }
+
```
Comment (ljharb) [inline lib/rules/no-object-type-as-default-prop.js:61]:
```suggestion
  properties.filter((prop) => prop.type === 'Property').forEach((prop) => {
```

File: lib/rules/no-object-type-as-default-prop.js [hunk 11]
```diff
@@ -0,0 +1,150 @@
...
+      context.report({
+        node: propDefaultValue,
+        messageId: MESSAGE_ID,
+        data: {
+          propName,
+          forbiddenType: 'regex literal'
+        }
+      });
+    } else if (propDefaultValueType === 'CallExpression'
+      && propDefaultValue.right.callee.type === 'Identifier'
```
Comment (ljharb) [inline lib/rules/no-object-type-as-default-prop.js:84]:
```suggestion
    } else if (
      propDefaultValueType === 'CallExpression'
      && propDefaultValue.right.callee.type === 'Identifier'
```

**Review by ljharb — CHANGES_REQUESTED** [review]
Overall, I think this looks good. After this pass, if my only feedback is docs prose, I'll tweak it myself prior to landing.

**Comment by fengkx** [comment]
Is there anything more to be done before this rule can be release? It looks very useful.  cc @ljharb

**Comment by ljharb** [comment]
@fengkx the review comment can be addressed either by @cyan33, or, by someone commenting a link to their branch (NOT another PR) that addresses them.

**Review by ljharb — CHANGES_REQUESTED** [review]
This rule should use the `Components.detect` utility rather than trying to implement its own component detection.

**Comment by ljharb** [comment]
@fengkx thanks! this PR has been freshly rebased, and I updated the test conventions as well. The only missing piece here is to use `Components.detect`. Please ping me when your branch link is updated and I'll be able to pull it in and rereview.

**Comment by fengkx** [comment]
[This branch](https://github.com/fengkx/eslint-plugin-react/tree/pr-2848-continue?rgh-link-date=2022-10-11T12%3A37%3A00Z) switched the component detection to use `Components.detect`. Because the `Component.detect` function no only check the function has a leading upper case letter but also return value( to be jsx or null), I change all test cases to return a `null`.

I can't find any doc about the `Components.detect` function. Should we mention this kind of utils in CONTRIBUTING.md ?

**Comment by ljharb** [comment]
Sure, it couldn't hurt to mention it in CONTRIBUTING.md.

Thanks, the added commit looks good; i'll pull it in shortly.

**Review by ljharb — APPROVED** [review]

**Comment by fengkx** [comment]
Could you release a new version for this? @ljharb

**Comment by ljharb** [comment]
When I’m prepared to field bug reports from the unreleased changes, i will.
